### DISCUSSION TITLE
Versioning event with complex type

### DISCUSSION THREAD

**Discussion by gaciv12** [akkadotnet/akka.net#5697] [Q&A] [answered]
Hi

I wanted to ask how do you solve problems with large objects (containing 6-7 fields) in Akka Persistance.

For example, code below:

```
record HouseActorModel (Guid Id, string Name, long Height, long width, HeatingInfoActorModel heatingInfo, ReadOnlyList<RoomActorModel>, ImmutableDictionary<Guid, ResidentActorModel>, ColorActorEnum Color)

abstract class HeatingInfoActorModel {
  HeatingTypeActorEnum HeatingType {get; set; }
}

class FloorHeatingActorModel: HeatingInfoActorModel
{
  int PipeUnderFloorLength {get; set; }
}

class CentralHeatingActorModel: HeatingInfoActorModel
{
  string ProviderName {get; set; }
}

class RoomActorModel
{
  int Size {get; set; }
  string Name {get; set; }
  bool HasInternet {get; set; }
}

class ResidentActorModel
{
  Guid Id {get; set; }
  string Name {get; set; }
  int Height {get; set; }
}

enum ColorActorEnum
{
 Black,
 White,
 Pink,
 ...
}
```

What should adding a new HouseActorModel look like?, with an eye to schema evolution for aka persistance.
I consider two approaches.
First:
Create a HouseActorModel object and pass it in Tell method: `houseManagerActor.Tell(AddHouse(new HouseActorModel(...))`
When saving the event to the database, also use this model, e.g. `Persist(new Envelope(new HouseActorModel (...)) => ...)`
The biggest disadvantage here is the amount of changes in case `ResidentActorModel` changes and I need `ResidentActorModelV2`. In this case I have to version all the above classes (more than 7 V2 classes).

The second solution is to replace the entire House model with simple types, e.g. `houseManagerActor.Tell (AddHouse (Id, Name, Height, width, ...ProblemHere...)`. It is much easier for me in the case of persist, because I am upgrading only events, not additionally objects in the event payload, but the problem is to expand complex objects into simple types. What about abstract class?

How do you deal with something like this? Do you change complex types to simple types, version multiple complex models, or maybe you have a different approach?

Thank you in advance :)

PS. I also asked on gitter too https://gitter.im/akkadotnet/akka.net?at=621cc1d50909252318003793, but no one answered to me :cry

**Comment by eaba** [comment]
Hello @gaciv12 I saw the question you asked on `gitter`, can we talk about what you need to accomplish so that I can understand you properly.

**Reply by gaciv12** [reply]
Yes, I need to persist `HouseActorModel` in Akka Persistance (SQLite), but versioning this payload of event is very hard because `HouseActorModel` type is very complex, so I don't want to version HouseActorModel and all types which `HouseActorModel` uses like `HeatingInfoActorModel`, `ResidentActorModel`, `RoomActorModel` etc and create `HouseActorModelV2`, `HeatingInfoActorModelV`2, `ResidentActorModelV2`, `RoomActorModelV2`, example:
`Persist(new Envelope(new HouseAddedV1(new HouseActorModelV1(...))) => ...)`
`Persist(new Envelope(new HouseAddedV2(new HouseActorModelV2(...))) => ...)`

 I want to version only event - example:
`Persist(new Envelope(new HouseAddedV1(... all properties of HouseActorModel ...)`
`Persist(new Envelope(new HouseAddedV2(... all properties of HouseActorModel with one renamed property ...)`

Do know what I mean?
Thank you in advance :)

**Comment by Aaronontheweb** [comment]
https://petabridge.com/cluster/lesson3 - I'll put together a YouTube video that explains how to do this in more detail, but basically you need to use custom serialization and apply https://aaronstannard.com/extend-only-design/ to your models.

**Reply by gaciv12** [reply]
Ok, but I don't have a problem with serialization. I will use EventAdapters. I have a problem with model of  versioning persist complex type (more code below)

**Reply by gaciv12** [reply]
What is the approximate time when you will record this video? Can you paste the link here after you've recorded it?

**Comment by gaciv12** [comment]
More code below

```
record HouseActorModel (Guid Id, string Name, long Height, long Width, HeatingInfoActorModel HatingInfo, ReadOnlyList<RoomActorModel> Rooms, ImmutableDictionary<Guid, ResidentActorModel> Residents, ColorActorEnum Color)

abstract class HeatingInfoActorModel {
  HeatingTypeActorEnum HeatingType {get; set; }
}

class FloorHeatingActorModel: HeatingInfoActorModel
{
  int PipeUnderFloorLength {get; set; }
}

class CentralHeatingActorModel: HeatingInfoActorModel
{
  string ProviderName {get; set; }
}

class RoomActorModel
{
  int Size {get; set; }
  string Name {get; set; }
  bool HasInternet {get; set; }
}

class ResidentActorModel
{
  Guid Id {get; set; }
  string Name {get; set; }
  int Height {get; set; }
}

ColorActorEnum
{
 Black,
 White,
 Pink,
 ...
}

HeatingTypeActorEnum
{
  Floor,
  Central
}

// ICommand and IEvent are my types for use only IEvent in Akka Persistance

public record AddHouse(HouseActorModel HouseActorModel) : ICommand
public record UpdateHouse(HouseActorModel HouseActorModel) : ICommand;

public record HouseAdded(HouseActorModel HouseActorModel) : IEvent;
public record HouseUpdated(HouseActorModel HouseActorModel) : IEvent;

private void Handle(AddHouse msg)
{
    this.Persist(new HouseAdded(msg.HouseActorModel), _ =>
    {
        this.Sender.Tell(Result.Success());
    });
}

private void Replay(HouseAdded houseAddedEvent)
{
    logger.LogInformation(houseAddedEvent.HouseActorModel);
}
    
    
private void Handle(UpdateHouse msg)
{
    this.Persist(new HouseUpdated(msg.HouseActorModel), evt =>
    {
        this.Sender.Tell(Result.Success());
    });
}

private void Replay(HouseUpdated houseUpdatedEvent)
{
    logger.LogInformation(houseUpdatedEvent.HouseActorModel);
}
```

1. Should I create all new model classes with V2 suffix for versioning (both top level and nested ones)
2. Should I add `CentralV2` in `HeatingTypeActorEnum`

But my main questions are: 
3. What should I do if I want change `string ProviderName` to `int ProviderCode` in `CentralHeatingActorModel?`
4. What should I do if I want change `string Name` to `string FullName` in `ResidentActorModel?`

I consider to change Reply methods from `Replay(HouseAdded houseAddedEvent)` to `Replay(Id, Name, Height ...`

**Reply by Aaronontheweb** [reply]
Going to break this up into two videos - have a script for the first, working on the code sample for it today.

**Reply by Aaronontheweb** [reply]
Video is live here: https://www.youtube.com/watch?v=oRll1Mzoyl4

**Comment by gaciv12** [comment]
Bump, help please

**Reply by eaba** [reply]
@gaciv12 morning! Have you tried this => `...basically you need to use custom serialization and apply https://aaronstannard.com/extend-only-design/ to your models.`?

**Reply by eaba** [reply]
`extend-only-design` means you should not change existing API, but create `overload`s (constructor, method) or make new parameters optional!

**Reply by gaciv12** [reply]
Yes, I tried this, but 
> 
> 3. What should I do if I want change `string ProviderName` to `int ProviderCode` in `CentralHeatingActorModel?` 
> 4. What should I do if I want change `string Name` to `string FullName` in `ResidentActorModel?`
> 
https://github.com/akkadotnet/akka.net/discussions/5697#discussioncomment-2280735

**Reply by eaba** [reply]
Probably you shouldn't, just add these a new field or property and then check for the existence of these new fields at the consuming side.

**Comment by Aaronontheweb** [comment, marked as answer]
Sorry for the delay on this, but my YouTube video explaining how to do this is finally finished: https://www.youtube.com/watch?v=oRll1Mzoyl4

It's about an hour long but covers the subject in detail, along with troubleshooting a live upgrade to existing event and state types.

**Reply by Aaronontheweb** [reply]
cc @gaciv12
